src/java/org/apache/cassandra/index/sai/utils/TypeUtil.java

```diff
@@ -0,0 +1,486 @@
+                                                          ComplexColumnData cellData,
+                                                          ColumnMetadata columnMetadata,
+                                                          IndexTarget.Type indexType,
+                                                          int nowInSecs)
+    {
+        if (cellData == null)
+            return null;
+
+        Stream<ByteBuffer> stream = StreamSupport.stream(cellData.spliterator(), false).filter(cell -> cell != null && cell.isLive(nowInSecs))
+                                                 .map(cell -> cellValue(columnMetadata, indexType, cell));
+
+        if (isInetAddress(validator))
+            stream = stream.sorted((c1, c2) -> compareInet(encodeInetAddress(c1), encodeInetAddress(c2)));
+
+        return stream.iterator();
+    }
+
+    public static Comparator<ByteBuffer> comparator(AbstractType<?> type)
+    {
+        // Override the comparator for BigInteger, frozen collections and composite types
+        if (isBigInteger(type) || isBigDecimal(type) || isCompositeOrFrozen(type))
+            return FastByteOperations::compareUnsigned;
+
+        return type;
+    }
+
+    private static ByteBuffer cellValue(ColumnMetadata columnMetadata, IndexTarget.Type indexType, Cell<?> cell)
+    {
+        if (columnMetadata.type.isCollection() && columnMetadata.type.isMultiCell())
+        {
+            switch (((CollectionType<?>) columnMetadata.type).kind)
+            {
+                case LIST:
+                    return cell.buffer();
+                case SET:
+                    return cell.path().get(0);
+                case MAP:
+                    switch (indexType)
+                    {
+                        case KEYS:
+                            return cell.path().get(0);
+                        case VALUES:
+                            return cell.buffer();
+                        case KEYS_AND_VALUES:
+                            return CompositeType.build(ByteBufferAccessor.instance, cell.path().get(0), cell.buffer());
+                    }
+            }
+        }
+        return cell.buffer();
+    }
+
+    /**
+     * Compares 2 InetAddress terms by ensuring that both addresses are represented as
+     * ipv6 addresses.
+     */
+    private static int compareInet(ByteBuffer b1, ByteBuffer b2)
+    {
+        assert isIPv6(b1) && isIPv6(b2);
+
+        return FastByteOperations.compareUnsigned(b1, b2);
+    }
+
+    private static boolean isIPv6(ByteBuffer address)
+    {
+        return address.remaining() == 16;
+    }
+
+    /**
+     * Encode a {@link InetAddress} into a fixed width 16 byte encoded value.
+     *
+     * The encoded value is byte comparable and prefix compressible.
+     *
+     * The encoding is done by converting ipv4 addresses to their ipv6 equivalent.
+     */
+    private static ByteBuffer encodeInetAddress(ByteBuffer value)
```

Comment:
Just to confirm, these fixed-width encodings wouldn't be strictly necessary for this patch, given we don't yet have the on-disk formats, correct?

Reply:
That's correct but, in the case of ip addresses it is used to normalise the address so we can handle ipv4 & ipv6 addresses more easily in the index.